crates/diesel_models/src/query/events.rs

```diff
@@ -222,4 +206,94 @@ impl Event {
         )
         .await
     }
+
+    fn apply_filters<T>(
+        mut query: T,
+        profile_id: Option<common_utils::id_type::ProfileId>,
+        from_to: (
+            dsl::created_at,
+            time::PrimitiveDateTime,
+            time::PrimitiveDateTime,
+        ),
+        limit: Option<i64>,
+        offset: Option<i64>,
+    ) -> T
+    where
+        T: diesel::query_dsl::methods::LimitDsl<Output = T>
+            + diesel::query_dsl::methods::OffsetDsl<Output = T>,
+        T: diesel::query_dsl::methods::FilterDsl<
+            diesel::dsl::GtEq<dsl::created_at, time::PrimitiveDateTime>,
+            Output = T,
+        >,
+        T: diesel::query_dsl::methods::FilterDsl<
+            diesel::dsl::LtEq<dsl::created_at, time::PrimitiveDateTime>,
+            Output = T,
+        >,
+        T: diesel::query_dsl::methods::FilterDsl<
+            diesel::dsl::Eq<dsl::business_profile_id, common_utils::id_type::ProfileId>,
+            Output = T,
+        >,
+    {
+        if let Some(profile_id) = profile_id {
+            query = query.filter(dsl::business_profile_id.eq(profile_id));
+        }
+
+        query = query
+            .filter(from_to.0.ge(from_to.1))
+            .filter(from_to.0.le(from_to.2));
+
+        if let Some(limit) = limit {
+            query = query.limit(limit);
+        }
+
+        if let Some(offset) = offset {
+            query = query.offset(offset);
+        }
+
+        query
+    }
+
+    pub async fn count_initial_attempts_by_constraints(
+        conn: &PgPooledConn,
+        merchant_id: &common_utils::id_type::MerchantId,
+        profile_id: Option<common_utils::id_type::ProfileId>,
+        created_after: time::PrimitiveDateTime,
+        created_before: time::PrimitiveDateTime,
+    ) -> StorageResult<i64> {
+        use async_bb8_diesel::AsyncRunQueryDsl;
+        use diesel::{debug_query, pg::Pg, QueryDsl};
+        use error_stack::ResultExt;
+        use router_env::logger;
+
+        use super::generics::db_metrics::{track_database_call, DatabaseOperation};
+        use crate::errors::DatabaseError;
+
+        let mut query = Self::table()
+            .count()
+            .filter(
+                dsl::event_id
+                    .nullable()
+                    .eq(dsl::initial_attempt_id) // Filter initial attempts only
+                    .and(dsl::merchant_id.eq(merchant_id.to_owned())),
+            )
+            .into_boxed();
+
+        query = Self::apply_filters(
+            query,
+            profile_id,
+            (dsl::created_at, created_after, created_before),
+            None,
+            None,
+        );
+
+        logger::debug!(query = %debug_query::<Pg, _>(&query).to_string());
+
+        track_database_call::<Self, _, _>(
+            query.get_result_async::<i64>(conn),
+            DatabaseOperation::Count,
+        )
+        .await
+        .change_context(DatabaseError::Others) // Query returns empty Vec when no records are found
```

Comment:
Can either update or remove the comment: This query would not be returning a `Vec` though.